forem/forem
# Add mention autocomplete to posts

## What type of PR is this? (check all applicable)


## Description

This PR adds the mention autocomplete functionality to the post editor. It also removes the use of the "preact-textarea-autosize" package in the post editor. After this PR we will be able to do a follow up 15 min fix to remove other usages, utilising the new hook instead.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/16
#13044
#13061  

## QA Instructions, Screenshots, Recordings

For testing, we'll need to verify that all of the core article form functionality continues to work as expected - including: preview, publish, edit existing draft, image drag and drop, image copy/paste.

We will need to test **both** versions of the editor and make sure that the text area height continues to expand to fit the content, as before. 

In addition to working as before, you should now be able to use the autocomplete functionality as in #13061 , by typing `@username`.

Video with V1 editor:

https://user-images.githubusercontent.com/20773163/112291377-f68f6b80-8c87-11eb-8aff-b0e20a969f30.mp4

Video with V2 editor:

https://user-images.githubusercontent.com/20773163/112291422-014a0080-8c88-11eb-8f83-62327736a1ab.mp4


### UI accessibility concerns?

Accessibility concerns are the same noted on #13061 and #13044:

> This refactor means that the whole textarea is now assigned the role "combobox". Screen reader users are immediately informed that a list of options will become associated with the input. An aria-live has been added to announce to screen reader users when the @mention search has started populating the list of options.

> It's important to note that axe flags an error for this component, as according to the WAI ARIA guidelines a combobox role should not apply to a textarea but a single line input. We have had a lot of discussions about how best to handle the screen reader use-case, and the implementation in this PR is very similar to the one GitHub is using for mentions. It makes sure that screen reader users can take advantage of the autocomplete functionality, despite it taking place in a textarea.

> I also don't believe this implementation will degrade the current editor/comment experience for screen reader users (when added there), since in all other respects the textarea continues to operate as before.

> In short, I think continuing in this direction is the best a11y experience we can offer to the best of our knowledge right now, but we should look in future to get this audited/tested by real users of assistive technology and be open to any feedback we receive.


## Added tests?

- [X] Yes - most of the autocomplete functionality has already been tested via the previous PR, but this PR adds some Cypress tests for mention functionality in posts

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      or in a [forem.dev](http://forem.dev) post


## [optional] What gif best describes this PR or how it makes you feel?

![a dog tries to reach a tennis ball on a sofa - text says 'almost there'](https://i.giphy.com/media/xUOxfh6ZM75efM3Bqo/giphy.webp)
